File: typescript/src/main/vsdk/toolkit/raycasting/bidirectionalRaytracing/DensityBuffer.ts
import { ColorRgb } from "../../common/color/ColorRgb";
import { Numeric } from "../../common/linealAlgebra/Numeric";
import { Vector2D } from "../../common/linealAlgebra/Vector2D";
import { ScreenBuffer } from "../../render/ScreenBuffer";
import { BidirectionalPathRaytracerConfig } from "./BidirectionalPathRaytracerConfig";
import { DensityHit } from "./DensityHit";
import { DensityHitList } from "./DensityHitList";
import { Kernel2D } from "./Kernel2D";

export class DensityBuffer {
  private static readonly DHA_X_RES = 50;
  private static readonly DHA_Y_RES = 50;

  private screenBuffer: ScreenBuffer;
  private baseConfig: BidirectionalPathRaytracerConfig;
  private xMinimum: number;
  private xMaximum: number;
  private yMinimum: number;
  private yMaximum: number;
  private hitGrid: DensityHitList[][];

  private xIndex(x: number): number {
    return globalThis.Math.min(
      globalThis.Math.trunc(DensityBuffer.DHA_X_RES * (x - this.xMinimum) / (this.xMaximum - this.xMinimum)),
      DensityBuffer.DHA_X_RES - 1
    );
  }

  private yIndex(y: number): number {
    return globalThis.Math.min(
      globalThis.Math.trunc(DensityBuffer.DHA_Y_RES * (y - this.yMinimum) / (this.yMaximum - this.yMinimum)),
      DensityBuffer.DHA_Y_RES - 1
    );
  }

  public constructor(screen: ScreenBuffer, paramBaseConfig: BidirectionalPathRaytracerConfig) {
    this.screenBuffer = screen;
    this.baseConfig = paramBaseConfig;

    this.xMinimum = this.screenBuffer.getScreenXMin();
    this.xMaximum = this.screenBuffer.getScreenXMax();
    this.yMinimum = this.screenBuffer.getScreenYMin();
    this.yMaximum = this.screenBuffer.getScreenYMax();

    this.hitGrid = new Array<DensityHitList[]>(DensityBuffer.DHA_X_RES);
    for (let i = 0; i < DensityBuffer.DHA_X_RES; i++) {
      this.hitGrid[i] = new Array<DensityHitList>(DensityBuffer.DHA_Y_RES);
      for (let j = 0; j < DensityBuffer.DHA_Y_RES; j++) {
        this.hitGrid[i][j] = new DensityHitList();
      }
    }

    process.stdout.write(
      "Density Buffer :\nXmin " + this.xMinimum + ", Ymin " + this.yMinimum + ", Xmax " + this.xMaximum + ", Ymax " + this.yMaximum + "\n"
    );
  }

  public add(x: number, y: number, color: ColorRgb): void {
    const factor = this.screenBuffer.getPixXSize() * this.screenBuffer.getPixYSize() * this.baseConfig.totalSamples;
    const tmpCol = new ColorRgb();

    if (color.average() > Numeric.EPSILON) {
      tmpCol.scaledCopy(factor, color);

      const hit = new DensityHit(x, y, tmpCol);

      this.hitGrid[this.xIndex(x)][this.yIndex(y)].add(hit);
    }
  }

  public reconstruct(): ScreenBuffer {
    const h = 8.0 * globalThis.Math.max(this.screenBuffer.getPixXSize(), this.screenBuffer.getPixYSize())
      / globalThis.Math.sqrt(this.baseConfig.samplesPerPixel);

    process.stdout.write("h = " + h + "\n");

    this.screenBuffer.scaleRadiance(0.0);

    const kernel = new Kernel2D();
    const center = new Vector2D();

    kernel.SetH(h);

    for (let i = 0; i < DensityBuffer.DHA_X_RES; i++) {
      for (let j = 0; j < DensityBuffer.DHA_Y_RES; j++) {
        const maxK = this.hitGrid[i][j].storedHits();

        for (let k = 0; k < maxK; k++) {
          const hit = this.hitGrid[i][j].get(k);

          center.x = hit.m_x;
          center.y = hit.m_y;

          kernel.cover(center, 1.0 / this.baseConfig.totalSamples, hit.color, this.screenBuffer);
        }
      }
    }

    return this.screenBuffer;
  }

  public reconstructVariable(dest: ScreenBuffer, baseSize = 4.0): ScreenBuffer {
    dest.scaleRadiance(0.0);

    const kernel = new Kernel2D();
    const center = new Vector2D();

    for (let i = 0; i < DensityBuffer.DHA_X_RES; i++) {
      for (let j = 0; j < DensityBuffer.DHA_Y_RES; j++) {
        const maxK = this.hitGrid[i][j].storedHits();

        for (let k = 0; k < maxK; k++) {
          const hit = this.hitGrid[i][j].get(k);

          center.x = hit.m_x;
          center.y = hit.m_y;

          kernel.varCover(
            center,
            hit.color,
            this.screenBuffer,
            dest,
            this.baseConfig.totalSamples,
            this.baseConfig.samplesPerPixel,
            baseSize
          );
        }
      }
    }

    return dest;
  }
}

File: java/base/src/main/java/vsdk/toolkit/raycasting/bidirectionalRaytracing/DensityBuffer.java
/**
Density estimation on screen
*/

package vsdk.toolkit.raycasting.bidirectionalRaytracing;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.common.color.ColorRgbMutable;
import vsdk.toolkit.common.linealAlgebra.Numeric;
import vsdk.toolkit.common.linealAlgebra.Vector2D;
import vsdk.toolkit.render.ScreenBuffer;

public class DensityBuffer {
    private static final int DHA_X_RES = 50;
    private static final int DHA_Y_RES = 50; // Subdivide image plane for efficient hit searches

    // A matching screen buffer is kept. This one will be filled in
    // by density estimation...
    private ScreenBuffer screenBuffer;
    private BidirectionalPathRaytracerConfig baseConfig;
    private float xMinimum; // Copy from screenBuffer
    private float xMaximum;
    private float yMinimum;
    private float yMaximum;
    private DensityHitList[][] hitGrid = new DensityHitList[DHA_X_RES][DHA_Y_RES];

    private int xIndex(float x) {
        return Math.min((int)(DensityBuffer.DHA_X_RES * (x - xMinimum) / (xMaximum - xMinimum)), DensityBuffer.DHA_X_RES - 1);
    }

    private int yIndex(float y) {
        return Math.min((int)(DensityBuffer.DHA_Y_RES * (y - yMinimum) / (yMaximum - yMinimum)), DensityBuffer.DHA_Y_RES - 1);
    }

    public DensityBuffer(ScreenBuffer screen, BidirectionalPathRaytracerConfig paramBaseConfig) {
        screenBuffer = screen;
        baseConfig = paramBaseConfig;

        xMinimum = screenBuffer.getScreenXMin();
        xMaximum = screenBuffer.getScreenXMax();
        yMinimum = screenBuffer.getScreenYMin();
        yMaximum = screenBuffer.getScreenYMax();

        for ( int i = 0; i < DHA_X_RES; i++ ) {
            for ( int j = 0; j < DHA_Y_RES; j++ ) {
                hitGrid[i][j] = new DensityHitList();
            }
        }

        System.out.printf("Density Buffer :\nXmin %f, Ymin %f, Xmax %f, Ymax %f\n",
            xMinimum, yMinimum, xMaximum, yMaximum);

    }

    /**
    Add a hit
    */
    public void add(float x, float y, ColorRgb color) {
        float factor = screenBuffer.getPixXSize() * screenBuffer.getPixYSize()
            * (float)baseConfig.totalSamples;
        ColorRgbMutable tmpCol = new ColorRgbMutable();

        if ( color.average() > Numeric.EPSILON ) {
            tmpCol.scaledCopy(factor, new ColorRgbMutable(color)); // Undo part of flux to rad factor

            DensityHit hit = new DensityHit(x, y, tmpCol.toImmutable());

            hitGrid[xIndex(x)][yIndex(y)].add(hit);
        }
    }

    /**
    Reconstruct the internal screen buffer using constant kernel width
    */
    public ScreenBuffer reconstruct() {
        // For all samples -> compute pixel coverage

        // Kernel size. Now spread over 3 pixels
        float h = (float)(8.0 * Math.max(screenBuffer.getPixXSize(), screenBuffer.getPixYSize())
            / Math.sqrt((float)baseConfig.samplesPerPixel));

        System.out.printf("h = %f\n", h);

        screenBuffer.scaleRadiance(0.0f); // Hack!

        int maxK;
        DensityHit hit;
        Kernel2D kernel = new Kernel2D();
        Vector2D center = new Vector2D();

        kernel.SetH(h);

        for ( int i = 0; i < DensityBuffer.DHA_X_RES; i++ ) {
            for ( int j = 0; j < DensityBuffer.DHA_Y_RES; j++ ) {
                maxK = hitGrid[i][j].storedHits();

                for ( int k = 0; k < maxK; k++ ) {
                    hit = hitGrid[i][j].get(k);

                    center.x = hit.m_x;
                    center.y = hit.m_y;

                    kernel.cover(center, 1.0f / (float)baseConfig.totalSamples, hit.color, screenBuffer);
                }
            }
        }

        return screenBuffer;
    }


    public ScreenBuffer reconstructVariable(ScreenBuffer dest, float baseSize) {
        // For all samples -> compute pixel coverage

        // Base Kernel size. Now spread over a number of pixels

        dest.scaleRadiance(0.0f); // Hack!

        int maxK;
        DensityHit hit;
        Kernel2D kernel = new Kernel2D();
        Vector2D center = new Vector2D();

        for ( int i = 0; i < DensityBuffer.DHA_X_RES; i++ ) {
            for ( int j = 0; j < DensityBuffer.DHA_Y_RES; j++ ) {
                maxK = hitGrid[i][j].storedHits();

                for ( int k = 0; k < maxK; k++ ) {
                    hit = hitGrid[i][j].get(k);

                    center.x = hit.m_x;
                    center.y = hit.m_y;

                    kernel.varCover(center, hit.color, screenBuffer, dest, (int)baseConfig.totalSamples,
                        baseConfig.samplesPerPixel, baseSize);
                }
            }
        }

        return dest;
    }

    public ScreenBuffer reconstructVariable(ScreenBuffer dest) {
        return reconstructVariable(dest, 4.0f);
    }
}
